skorch/net.py

```diff
@@ -487,9 +487,97 @@ def validation_step(self, Xi, yi, **fit_params):
             'y_pred': y_pred,
             }
 
+    class TrainStepPrepare:
+        """Callable passed to an optimizer. Computes y_pred, loss value. and
+        updates gradients of the loss function w.r.t. model parameters. At the
+        same time, records intermediate 'y_pred' and 'loss' values.
+
+        Values from the first evaluation are returned. In case of some
+        optimizers, e.g. LBFGS, it is called multiple times, as the loss
+        function is evalueted multiple times per optimizer step.
+
+        In such cases, if all loss function values are needed,
+        `store_step_values` and `get_step_values` should be overridden in
+        a subclass.
+        """
+        def __init__(self, net, Xi, yi, **fit_params):
+            """Prepare the callable by storing arguments
+
+            Parameters
+            ----------
+            net : skorch.NeuralNet
+              Net being trained
+
+            Xi : input data
+              A batch of the input data.
+
+            yi : target data
+              A batch of the target data.
+
+            **fit_params : dict
+              Additional parameters passed to the ``forward`` method of
+              the module and to the train_split call.
+
+            """
+            self.step_values = None
+            self.net = net
+            self.Xi = Xi
+            self.yi = yi
+            self.fit_params = fit_params
+
+        def __call__(self):
+            """Compute y_pred, loss value, and update net's gradients.
+
+            The module is set to be in train mode (e.g. dropout is
+            applied).
+            """
+            self.net.module_.train()
+            self.net.optimizer_.zero_grad()
+            y_pred = self.net.infer(self.Xi, **self.fit_params)
+            loss = self.net.get_loss(y_pred, self.yi, X=self.Xi, training=True)
+            loss.backward()
+
+            self.net.notify(
+                'on_grad_computed',
+                named_parameters=list(self.net.module_.named_parameters())
+            )
+
+            step_values = {
+                'loss': loss,
+                'y_pred': y_pred,
+                }
+            self.store_step_values(step_values)
+            return loss
```

Comment:
I get the point about not breaking a pattern of how this library is designed, I've tried to accommodate it in the new commit.

Since `train_step_single` has to have a reference to the accumulator anyway (to call `store_step_values` - it was possible to simplify the accumulator class even further.


About naming: I would find it really confusing if `train_step_single` would not actually update the net's parameters. Is `train_step_gradient_calc` okay?